File: frontend-design/scripts/generate_chart_brief.py
#!/usr/bin/env python3
"""Generate chart/data-visualization guidance.

Usage:
  python scripts/generate_chart_brief.py comparison revenue categories
  python scripts/generate_chart_brief.py realtime monitoring latency
"""

from __future__ import annotations

import argparse
from _catalog import rank_csv, pick, md_bullet


def main() -> int:
    parser = argparse.ArgumentParser(description="Generate chart brief")
    parser.add_argument("query", nargs="+")
    parser.add_argument("-n", "--limit", type=int, default=3)
    args = parser.parse_args()
    query = " ".join(args.query)
    rows = rank_csv("chart-selection.csv", query, args.limit)
    slide_rows = rank_csv("slide-charts.csv", query, min(args.limit, 2))
    print(f"# Chart Brief: {query}\n")
    if rows:
        print("## Chart Candidates\n")
        for row in rows:
            print(f"### {pick(row, 'Best Chart Type')}")
            for label, key in [
                ("Data type", "Data Type"), ("Secondary", "Secondary Options"),
                ("Use when", "When to Use"), ("Avoid when", "When NOT to Use"),
                ("Volume", "Data Volume Threshold"), ("Color", "Color Guidance"),
                ("Accessibility", "Accessibility Grade"), ("Interaction", "Interaction Pattern"),
            ]:
                value = pick(row, key)
                if value:
                    print(md_bullet(label, value))
            print("")
    if slide_rows:
        print("## Presentation Chart Fit\n")
        for row in slide_rows:
            print(f"### {pick(row, 'chart_type')}")
            print(md_bullet("Best for", pick(row, "best_for")))
            print(md_bullet("Data type", pick(row, "data_type")))
            print(md_bullet("Use", pick(row, "when_to_use")))
            print(md_bullet("Avoid", pick(row, "when_to_avoid")))
            print(md_bullet("Max categories", pick(row, "max_categories")))
            print("")
    print("## Rules\n")
    print("- Start from the question the user must answer.")
    print("- Label important values directly where possible.")
    print("- Do not rely on color alone; use text, position, shape, or icons.")
    print("- Preserve truthful scales and label truncation/log scales explicitly.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend-design/scripts/generate_chart_brief.mjs
#!/usr/bin/env node
/** Generate chart and data-visualization guidance. */

import { pathToFileURL } from "node:url";
import { mdBullet, pick, rankCsv } from "./_catalog.mjs";

const CHART_FIELDS = [
  ["Data type", "Data Type"], ["Secondary", "Secondary Options"], ["Use when", "When to Use"],
  ["Avoid when", "When NOT to Use"], ["Volume", "Data Volume Threshold"], ["Color", "Color Guidance"],
  ["Accessibility", "Accessibility Grade"], ["Interaction", "Interaction Pattern"],
];

function parseArgs(args) {
  let limit = 3;
  const query = [];
  for (let index = 0; index < args.length; index += 1) {
    if (args[index] === "-n" || args[index] === "--limit") {
      limit = Number.parseInt(args[index + 1], 10);
      index += 1;
    } else query.push(args[index]);
  }
  if (query.length === 0 || !Number.isInteger(limit)) throw new Error("Usage: generate_chart_brief.mjs <query> [-n <limit>]");
  return { query: query.join(" "), limit };
}

export function main(args = process.argv.slice(2)) {
  let options;
  try {
    options = parseArgs(args);
  } catch (error) {
    console.error(error.message);
    return 2;
  }
  const charts = rankCsv("chart-selection.csv", options.query, options.limit);
  const slides = rankCsv("slide-charts.csv", options.query, Math.min(options.limit, 2));
  const lines = [`# Chart Brief: ${options.query}`, ""];
  if (charts.length > 0) {
    lines.push("## Chart Candidates", "");
    for (const row of charts) {
      lines.push(`### ${pick(row, "Best Chart Type")}`);
      for (const [label, key] of CHART_FIELDS) {
        const value = pick(row, key);
        if (value) lines.push(mdBullet(label, value));
      }
      lines.push("");
    }
  }
  if (slides.length > 0) {
    lines.push("## Presentation Chart Fit", "");
    for (const row of slides) {
      lines.push(
        `### ${pick(row, "chart_type")}`,
        mdBullet("Best for", pick(row, "best_for")),
        mdBullet("Data type", pick(row, "data_type")),
        mdBullet("Use", pick(row, "when_to_use")),
        mdBullet("Avoid", pick(row, "when_to_avoid")),
        mdBullet("Max categories", pick(row, "max_categories")),
        "",
      );
    }
  }
  lines.push(
    "## Rules", "",
    "- Start from the question the user must answer.",
    "- Label important values directly where possible.",
    "- Do not rely on color alone; use text, position, shape, or icons.",
    "- Preserve truthful scales and label truncation/log scales explicitly.",
  );
  process.stdout.write(`${lines.join("\n")}\n`);
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) process.exitCode = main();
